python/ccf/clients.py

```diff
@@ -575,33 +598,40 @@ def call(
         :param dict headers: HTTP request headers (optional).
         :param bool signed: Sign request with client private key.
         :param int timeout: Maximum time to wait for a response before giving up.
+        :param list log_capture: Rather than emit to default handler, capture log lines to list (optional).
 
         :return: :py:class:`ccf.clients.Response`
         """
         if not path.startswith("/"):
             raise ValueError(f"URL path '{path}' is invalid, must start with /")
 
+        logs: List[str] = []
+
         if self.is_connected:
-            return self._direct_call(path, body, http_verb, headers, signed, timeout)
+            r = self._call(path, body, http_verb, headers, signed, timeout, logs)
+            flush_info(logs, log_capture, 2)
+            return r
 
         end_time = time.time() + self.connection_timeout
         while True:
             try:
-                response = self._direct_call(
-                    path, body, http_verb, headers, signed, timeout
+                logs = []
+                response = self._call(
+                    path, body, http_verb, headers, signed, timeout, logs
                 )
                 # Only the first request gets this timeout logic - future calls
-                # call _direct_call
+                # call _call
                 self.is_connected = True
+                flush_info(logs, log_capture, 2)
                 return response
             except (CCFConnectionException, TimeoutError) as e:
                 # If the initial connection fails (e.g. due to node certificate
                 # not yet being endorsed by the network) sleep briefly and try again
                 if time.time() > end_time:
+                    flush_info(logs, log_capture, 2)
```

Comment:
Here and in other calls, the intention is to hold off logging until the last iteration essentially, so as not to print retries. There's at least two things it could do better:

1. mangling timestamps to keep them correct, right now the request stamp is essentially the same as the response stamp. So if the last request took forever, that's not reported accurately.
2. reporting a retry count "After X retries...". This isn't super straightforward because these things chain up, and then you have to wonder if you report the retries of one step below or all the way down.

Reply:
"the request stamp is essentially the same as the response stamp. So if the last request took forever, that's not reported accurately"

This seems like a big loss. Can we fix it?

Reply:
I may have found a way to do this.

Reply:
No I haven't. I'm going to give up for now, because I think I've already spent too much time on it.